### PR TITLE
First Commit. Update readme. Flowchart in img folder

### PR DESCRIPTION
**Pull request by mingyikoh**
Readme is updated. 
Flowchart has two versions, one has better formatting. Delete other one in next commit.
Start work on coding game logic.

### CODE REVIEW

File: assets/js/script.js [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+
+// END OF GENERAL GAME FUNCTIONS -----------------------
+// SPAWN FUNCTIONS -------------------------------------
+
+  $gameScreen = $('.gameScreen')
+  var spawnList = []
+  var counter = 1
+
+  // constructor for all spawns
+  function Spawn(life, damageWhenExpire, effectHealth, effectAmmo, effectGrenade, timeToExpire, counterLink){
```
Comment (primaulia) [inline assets/js/script.js]: Why are we not using a `class` method here?

File: assets/js/script.js [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+
+  // Spawn stats (life, damageWhenExpire, effectHealth, effectAmmo, effectGrenade, timeToExpire, counterLink)
+  function spawnEnemyEasy () {
+    var enemyEasy = new Spawn(1, 1, 0, 0, 0, 3, 's'+counter)
+    spawnList.push(enemyEasy)
+    var $spawn = $('<div>')
+    $spawn.attr('id', "s"+counter)
+    $spawn.addClass('spawn enemyEasy')
+    $spawn.css({
+      "left": Math.floor(Math.random() * 1000),
```
Comment (primaulia) [inline assets/js/script.js]: Why don't you randomize the position in the constructor instead?

**Comment by primaulia** [comment]
Project Workflow: 3 / 5
Technical Requirements: 4 / 5
Creativity: 5 / 5
Code Quality: 3 / 5
Problem Solving: 4 / 5
Delivery: 5 / 5
Professional Skill: 4 / 5

# Glow
- Good delivery, nice attempt to invite your classmates to play the game
- Very fun game, I think you put a good thought into it in terms of design and interaction
- Good flowchart, it helps me to understand your code flow better

# Grow
- Without flowchart, your code is actually quite hard to follow, try to separate the code in a more logical group next time. Show your codes to your friend, ask them whether they can follow through
- There are various repetitions in your code that can be refactored

# Things to look out for
- Function flow
- Separation of concern between design / dom manipulation and game play update

File: assets/js/script.js [hunk 3]
```diff
@@ -0,0 +1,471 @@
+$(function() {
+
```
Comment (alexkimin) [inline assets/js/script.js:2]: If possible, declare variables first at the top of your scope.

File: assets/js/script.js [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+// SCREEN BUTTONS
+  function generalStartFunctions() {
+    $gameOverlay = $('.gameOverlay')
+    $gameOverlay.css({
+      'display': 'none'
+    })
+    document.getElementsByClassName('gogo')[0].volume = 0.6;
```
Comment (alexkimin) [inline assets/js/script.js]: You may replace this line `$('.gogo.')[0].volume = 0.6`. Why does that works but `$('.gogo').volume` doesn't work?

File: assets/js/script.js [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+    document.getElementsByClassName('gogo')[0].play()
+    updatePlayerStats()
+    countDown()
+  }
+
+  $('.easyStartBtn').on('click', () => {
```
Comment (alexkimin) [inline assets/js/script.js]: You may move your eventListeners after the declarations of functions.

File: assets/js/script.js [hunk 6]
```diff
@@ -0,0 +1,471 @@
...
+    ammo: 25,
+    grenade: 3,
+  }
+  var timeCount = 120
+  var playerStatsInterval
+  var countDownInterval
+
+  function updatePlayerStats () {
+    playerStatsInterval = setInterval(() => {
+      var $playerHealth = $('.playerHealth')
```
Comment (alexkimin) [inline assets/js/script.js]: What will be the difference if you put these line(100 - 103) in outer scope of function `updatePlayerStats`? What will be the advantage if we reduce the number of DOM accessing?

File: assets/js/script.js [hunk 7]
```diff
@@ -0,0 +1,471 @@
...
+
+  function countDown() {
+    countDownInterval = setInterval( () => {
+      var $timer = $('.timer')
+      timeCount = timeCount - 1
+      $timer.text('Timer: ' +timeCount+ ' seconds')
+      checkEnemyExpire()
+      checkVictory()
+      checkLoss()
+    }
```
Comment (alexkimin) [inline assets/js/script.js:44]: `}, 1000)`

File: assets/js/script.js [hunk 8]
```diff
@@ -0,0 +1,471 @@
...
+    , 1000)
+  }
+
+  function checkEnemyExpire() {
+    for(var i = spawnList.length -1; i > -1; i --){
+      spawnList[i].timeToExpire = spawnList[i].timeToExpire -1
+      if(spawnList[i].timeToExpire <0){
+        if(spawnList[i].damageWhenExpire > 0){
+          playerStats.health = playerStats.health - spawnList[i].damageWhenExpire
+          document.getElementsByClassName('playerDamage')[0].volume = 0.8
```
Comment (alexkimin) [inline assets/js/script.js:65]: How can we reduce DOM accessing?

File: assets/js/script.js [hunk 9]
```diff
@@ -0,0 +1,471 @@
...
+        spawnList.splice(i, 1)
+      }
+    }
+  }
+
+  function checkEnemyHealth () {
+    for (var i = spawnList.length -1; i > -1; i --) {
+      if(spawnList[i].life <= 0) {
+        $('div').remove('#'+spawnList[i].counterLink)
+        if (spawnList[i].effectHealth === -1) playerStats.health = playerStats.health -1
```
Comment (alexkimin) [inline assets/js/script.js]: you can do this `playerStats.health -= 1`

File: assets/js/script.js [hunk 10]
```diff
@@ -0,0 +1,471 @@
...
+
+  function spawnIntervalGrenade (interval) {
+    setInterval(spawnGrenadeRefill, interval)
+  }
+
+// END OF SPAWN FUNCTIONS -----------------------------
+// EVENTS ---------------------------------------------
+
+// 3 easy, ammo refill
+  function event1(time) {
```
Comment (alexkimin) [inline assets/js/script.js]: You may use better names for functions. :)

**Comment by alexkimin** [comment]
Project Workflow: 4 / 5
Technical Requirements: 4 / 5
Creativity: 4 / 5
Code Quality: 3 / 5
Problem Solving: 4 / 5

## Glow

Very fun game with sound and well-balanced modes.
Good flowchart.

## Grow

Naming convention on some functions can be improved.
Reducing DOM accessing if possible.

## Things to look out for

ES6 `Class`
[https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes](url)

File: assets/js/script.js (hunk 6 above)
Comment (mingyikoh) [reply assets/js/script.js]: Ah I see, putting the variables in the set interval scope will make unnecessary DOM accessing over and over again.

File: assets/js/script.js (hunk 2 above)
Comment (mingyikoh) [reply assets/js/script.js]: My intention was to separate the functions that are HTML/CSS from the JS side of things.

For my constructor I only wanted to include 'game effect stats' that had nothing to do with the HTML and CSS side of things. This is mainly because I was hoping to add more variety to enemy stats in the future, for example adding scores, or making boss enemies that have more life and stay for longer but deal set damage every few seconds. I was worried adding CSS inside the constructor would make it look too messy.

For the setting of ID, Class, HTML and CSS side of things, I was hoping to chunk them in a separate function. I realise that it resulted in a lot of repetition for my spawnEnemyEasy(), spawnEnemyNormal() and all the rest. My solution would be to create a separate function named spawnDiv() which would group the repetitions and replace the repetitions with this function. This would allow the elimination of repetition and still keep the constructor clean imo.
